pkg/tarfs/tarfs.go

```diff
@@ -305,83 +305,86 @@ func (t *Manager) getImageBlobInfo(metaFilePath string) (string, error) {
-			return
+			return errors.Wrapf(err, "can not found status object for snapshot %s after prepare", snapshotID)
 		}
-		defer st.mutex.Unlock()
+		defer func() {
+			if st.wg != nil {
+				st.wg.Done()
+				st.wg = nil
+			}
+			st.mutex.Unlock()
+		}()
 
 		st.blobID = layerBlobID
 		if err != nil {
-			log.L.WithError(err).Errorf(msg)
 			st.status = TarfsStatusFailed
 		} else {
 			st.status = TarfsStatusReady
 		}
-		log.L.Infof(msg)
-	}
 
-	keyChain, err := auth.GetKeyChainByRef(ref, nil)
-	if err != nil {
-		epilog(err, "create key chain for connection")
-		return err
-	}
-	remote := remote.New(keyChain, t.insecure)
-	rc, _, err := t.getBlobStream(ctx, remote, ref, layerDigest)
-	if err != nil && remote.RetryWithPlainHTTP(ref, err) {
-		rc, _, err = t.getBlobStream(ctx, remote, ref, layerDigest)
-	}
-	if err != nil {
-		epilog(err, "get blob stream for layer")
-		return errors.Wrapf(err, "get blob stream by digest")
+		return errors.Wrapf(err, msg)
 	}
 
-	go func() {
-		defer wg.Done()
+	process := func(rc io.ReadCloser, remote *remote.Remote) error {
 		defer rc.Close()
 
 		ds, err := compression.DecompressStream(rc)
 		if err != nil {
-			epilog(err, "unpack layer blob stream for tarfs")
-			return
+			return epilog(err, "unpack layer blob stream for tarfs")
 		}
 		defer ds.Close()
 
 		if t.validateDiffID {
 			diffID, err := t.getBlobDiffID(ctx, remote, ref, manifestDigest, layerDigest)
 			if err != nil {
-				epilog(err, "get layer diffID")
-				return
+				return epilog(err, "get layer diffID")
 			}
 			digester := digest.Canonical.Digester()
 			dr := io.TeeReader(ds, digester.Hash())
 			err = t.generateBootstrap(dr, snapshotID, layerBlobID, upperDirPath)
 			switch {
 			case err != nil && !errdefs.IsAlreadyExists(err):
-				epilog(err, "generate tarfs from image layer blob")
+				return epilog(err, "generate tarfs from image layer blob")
 			case err == nil && digester.Digest() != diffID:
-				epilog(err, "image layer diffID does not match")
+				return epilog(err, "image layer diffID does not match")
 			default:
-				msg := fmt.Sprintf("nydus tarfs for snapshot %s is ready, digest %s", snapshotID, digester.Digest())
-				epilog(nil, msg)
+				msg := fmt.Sprintf("Nydus tarfs for snapshot %s is ready, digest %s", snapshotID, digester.Digest())
+				return epilog(nil, msg)
 			}
 		} else {
 			err = t.generateBootstrap(ds, snapshotID, layerBlobID, upperDirPath)
 			if err != nil && !errdefs.IsAlreadyExists(err) {
-				epilog(err, "generate tarfs data from image layer blob")
+				return epilog(err, "generate tarfs data from image layer blob")
 			} else {
-				msg := fmt.Sprintf("nydus tarfs for snapshot %s is ready", snapshotID)
-				epilog(nil, msg)
+				msg := fmt.Sprintf("Nydus tarfs for snapshot %s is ready", snapshotID)
+				return epilog(nil, msg)
 			}
 		}
-	}()
+	}
+
+	keyChain, err := auth.GetKeyChainByRef(ref, nil)
+	if err != nil {
+		epilog(err, "create key chain for connection")
```

Comment:
It seems the `epilog(err, "create key chain for connection")` should be the value of `return`?

Reply:
fixed.